src/Infrastructure.VS.UnitTests/VsUIServiceOperationTests.cs

```diff
@@ -0,0 +1,319 @@
+            var calls = new List<string>();
+
+            var serviceProvider = new Mock<IServiceProvider>();
+            serviceProvider.Setup(x => x.GetService(typeof(S)))
+                .Callback(() => calls.Add("get service"));
+
+            // The method does the setup for the sync and async tests, so we'll set up both
+            // RunOnUIThread and RunOnUIThreadAsync
+            var threadHandling = new Mock<IThreadHandling>();
+            threadHandling.Setup(x => x.RunOnUIThread(It.IsAny<Action>()))
+                .Callback<Action>(x =>
+                {
+                    calls.Add("change thread - sync");
+                    x();
+                });
+
+            threadHandling.Setup(x => x.RunOnUIThreadAsync(It.IsAny<Action>()))
+                .Callback<Action>(x =>
+                {
+                    calls.Add("change thread - async");
+                    x();
+                });
+
+            var testSubject = CreateTestSubject(serviceProvider.Object, threadHandling.Object);
+
+            return (calls, testSubject);
+        }
+
+        #endregion
+
+        #region Func methods return results
+
+        [TestMethod]
+        public void Execute_Func_ReturnsValueFromCallback()
+        {
+            (var callback, var testSubject) = SetupUpFuncReturnsValueTests("expected sync result");
+
+            // Act
+            var result = testSubject.Execute<SMyInterface, IMyInterface, string>(callback.Object);
+
+            result.Should().Be("expected sync result");
+            callback.Verify(x => x.Invoke(It.IsAny<IMyInterface>()), Times.Once);
+        }
+
+        [TestMethod]
+        public async Task ExecuteAsync_Func_ReturnsValueFromCallback()
+        {
+            (var callback, var testSubject) = SetupUpFuncReturnsValueTests("expected async result");
+
+            // Act
+            var result = await testSubject.ExecuteAsync<SMyInterface, IMyInterface, string>(callback.Object);
+
+            result.Should().Be("expected async result");
+            callback.Verify(x => x.Invoke(It.IsAny<IMyInterface>()), Times.Once);
+        }
+
+        private static (Mock<Func<IMyInterface, string>> callback, VsUIServiceOperation vsServiceOperation) SetupUpFuncReturnsValueTests(string callbackResultToReturn)
+        {
+            var service = Mock.Of<IMyInterface>();
+            var serviceProvider = CreateConfiguredServiceProvider<SMyInterface>(service);
+            var callback = new Mock<Func<IMyInterface, string>>();
+            callback.Setup(x => x.Invoke(service)).Returns(callbackResultToReturn);
+
+            var testSubject = CreateTestSubject(serviceProvider.Object);
+
+            return (callback, testSubject);
+        }
+
+        #endregion
+
+        #region Multiple calls, same service tests
+
+        // NOTE - we're only testing this scenario for Execute(Action)
+        // Adding tests for all of the other methods doesn't add enough value to justify the effort at this point.
```

Comment:
As discussed offline, I couldn't work out an easy way to avoid a lot of duplication in this test and the next one, so there are only tests for one entry point.